src/pkg/services/m365/api/item_pager.go

```diff
@@ -27,6 +27,57 @@ type DeltaUpdate struct {
 	Reset bool
 }
 
+type NextPager[T any] interface {
+	NextPage() (items []T, reset, done bool)
+}
+
+type nextPage[T any] struct {
+	items []T
+	// reset is only true on the iteration where the delta pager's Reset()
+	// is called.  Callers can use it to reset any data aggregation they
+	// currently use.  After that loop, it will be false again, though the
+	// DeltaUpdate will still contain the expected value.
+	reset bool
+}
+
+type NextPageResulter[T any] interface {
+	NextPager[T]
+
+	Results() (DeltaUpdate, error)
+}
+
+var _ NextPageResulter[any] = &nextPageResults[any]{}
+
+type nextPageResults[T any] struct {
+	pages chan nextPage[T]
+	du    DeltaUpdate
+	err   error
+}
+
+func (npr *nextPageResults[T]) NextPage() ([]T, bool, bool) {
+	if npr.pages == nil {
+		return nil, false, true
+	}
+
+	np := <-npr.pages
+
+	return np.items, np.reset, false
+}
+
+func (npr *nextPageResults[T]) Results() (DeltaUpdate, error) {
+	// if the pager hasn't closed yet, drain out the pages iterator
+	// to avoid leaking routines, and to ensure we get results.
+	for npr.pages != nil {
+		<-npr.pages
+	}
```

Comment:
nit: could you please switch this to range directly over the channel? I think the code should just call `close(npr.pages)` when it's done instead of having a separate close method (unless there's a high probability of needing to do other work)

If it doesn't just call close on the channel then please somehow synchronize access with a mutex. With the current code it's possible to get an NPE if `close()` is called between the if-check and attempting to read from the channel

Reply:
Oh, huh, I forgot the close in close().  Yay for that.  Thanks for the catch.